### ISSUE TITLE
[Mac only] The app takes a long time to quit and apply the update after calling quitAndInstall

### ISSUE THREAD

**Issue by DamonYu6** [electron-userland/electron-builder#8651]
* **Electron-Builder Version**: 23.3.0

* **Node Version**: 18.17.1 (integrated within electron@27.2.0)

* **Electron Version**: 27.2.0
* **Electron Type (current, beta, nightly)**: current

* **Electron-Updater Version**: 6.3.4

* **Target**: Mac (Apple Silicon / Intel)

Issue: After upgrading our Electron app from Electron 23 to Electron 27, we sometimes find that after calling the quitAndInstall API, the app takes over 10 minutes to quit and apply the update.

Expected Behavior: The app should quit and apply the update immediately after calling the quitAndInstall API.

Analysis: By adding extra debug information, we found that after the app receives the downloaded event dispatched by electron-updater, it takes a long time before receiving the downloaded event from nativeUpdater. It’s currently unclear why the Squirrel server takes more than 10 minutes to pipe the update.zip file.

Solution: Upon reviewing the code, I discovered https://github.com/electron-userland/electron-builder/pull/5271, which altered the timing of the updater-downloaded event. Now, the updater dispatches the updater-downloaded event to the app layer as soon as the zip file is downloaded, but in reality, quitAndInstall only takes immediate effect when the updater-downloaded event from nativeUpdater is triggered. Therefore, I believe we should align the dispatch timing of the update-downloaded event with nativeUpdater by reverting the changes related to the dispatch timing in [PR 5271](https://github.com/electron-userland/electron-builder/pull/5271). This should also resolve the issue we’re currently encountering.

**Comment by mmaietta** [comment]
> It’s currently unclear why the Squirrel server takes more than 10 minutes to pipe the update.zip file.

I feel like this would be the next steps in terms of investigation. I wouldn't expect a localhost server to be noticably limited by bandwidth.

I'm hesitant to move forward with this change. From what I understand, the reason why the update-downloaded event happens after file download and not via the nativeUpdater is to allow this functionality in https://github.com/electron-userland/electron-builder/pull/5271
> to be able to do an additional cryptographic check on downloaded files and conditionally abort the update on a failed verification.

If I understand your need correctly, I think you can attach to the native event to have separate logic instead; I think you can use some code like below. (needs the `as any` since `nativeUpdater` is currently private on MacUpdater)
```
if (autoUpdater instanceof MacUpdater) {
   (autoUpdater as any).nativeUpdater.on("update-downloaded", () => { ... });
}
```

**Comment by DamonYu6** [comment]
Oh, I almost forgot this change would disrupt users who have set **autoInstallOnAppQuit = false** and rely on the downloaded event for update validation.

**Comment by mmaietta** [comment]
Right, which is why I don't want to proceed with the proposed change as that's a pre-existing security approach in a highly delicate area of the codebase. That's why I provided a workaround for you to tap into the native `electron-updater` supplied by electron.

Going to close this issue and related PR for now. Happy to continue discussing this topic though or assisting you in getting a workaround set up

**Comment by DamonYu6** [comment]
@mmaietta @beyondkmp We recently reproduced this issue, and by adding some logs to Squirrel.Mac, we found that the bottleneck is the [unzip archive step](https://github.com/Squirrel/Squirrel.Mac/blob/0e5d146ba13101a1302d59ea6e6e0b3cace4ae38/Squirrel/SQRLZipArchiver.m#L113) of the update package (Squirrel.Mac verifies the signature and unzips the package after downloading the .zip file, and only then triggers the update-downloaded event for Electron).

From the logs, we can see that the unzip process takes several minutes. Although we are not sure why unzipping the archive takes so long, we observed in Activity Monitor that despite the system having low CPU and memory usage, the "ditto" process still takes several minutes to complete the unzip operation.

Because our app relies on Electron Updater’s "downloaded" event to show a "ready for update" UI to the user, this causes many users to report that nothing happens after clicking "update" (the expected behavior is for the app to exit, upgrade, and then relaunch). In reality, the app only exits itself several minutes later.

I suggest adding a new event called "update-ready" in MacUpdater. This event would be dispatched externally after Electron’s "downloaded" event is triggered. I would be happy to submit a PR if you guys agree with it.

<img width="1278" height="743" alt="Image" src="https://github.com/user-attachments/assets/c890c284-ac16-4442-8960-2d6de4499709" />

**Comment by beyondkmp** [comment]
Please fire a pr.

**Comment by mmaietta** [comment]
Oh bugger, it's part of Squirrel that is the limitation. I am also looking at simplifying the mac auto-updater to not require a localhost server to serve to the native auto-updater and instead provide a file URL via `file://` protocol. Just haven't had a chance to investigate further, but the blackbox e2e tests now have coverage of that flow.
